openwisp_firmware_upgrader/static/firmware-upgrader/js/batch-upgrade-progress.js

```diff
@@ -0,0 +1,433 @@
+  let $ = django.jQuery;
+  let mainProgressElement = $(".batch-main-progress");
+  if (mainProgressElement.length > 0) {
+    let progressPercentage =
+      data.total > 0 ? Math.round((data.completed / data.total) * 100) : 0;
+    let statusClass = (data.status || "").replace(/\s+/g, "-");
+    let showPercentageText = true;
+
+    if (data.status === "success") {
+      progressPercentage = 100;
+      statusClass = "completed-successfully";
+      showPercentageText = true;
+    } else if (data.status === "cancelled") {
+      progressPercentage = 100;
+      statusClass = "cancelled";
+      showPercentageText = false;
+    } else if (data.status === "failed") {
+      let successfulOpsCount = $("#result_list tbody tr").filter(function () {
+        let statusText = $(this).find(".status-cell .status-content").text().trim();
+        return statusText === "success" || statusText === "completed successfully";
+      }).length;
+
+      // Also check individual operation containers for success
+      if (successfulOpsCount === 0) {
+        $("#result_list tbody tr").each(function () {
+          let statusContainer = $(this).find(".upgrade-status-container");
+          if (
+            statusContainer.length &&
+            statusContainer.find(".upgrade-progress-fill.success").length
+          ) {
+            successfulOpsCount++;
+          }
+        });
+      }
+      if (successfulOpsCount > 0) {
+        // Some operations succeeded - partial success (orange)
+        progressPercentage = 100;
+        statusClass = "partial-success";
+        showPercentageText = false;
+      } else {
+        // All operations failed - total failure (red)
+        progressPercentage = 100;
+        statusClass = "failed";
+        showPercentageText = false;
+      }
+    }
+
+    let progressHtml = `
+      <div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill ${statusClass}" style="width: ${progressPercentage}%"></div>
+      </div>
+    `;
+    if (showPercentageText) {
+      progressHtml += `<span class="upgrade-progress-text">${progressPercentage}%</span>`;
+    }
+
+    mainProgressElement.html(progressHtml);
+  }
+
+  // Update completion information in the admin form if available
+  if (data.total && data.completed) {
+    let completedInfo = $(".field-completed .readonly");
+    if (completedInfo.length > 0) {
+      completedInfo.text(`${data.completed} out of ${data.total}`);
+    }
+  }
+  let statusField = $(".field-status .readonly");
+  if (statusField.length > 0 && data.status) {
+    let displayStatus = data.status;
+    if (data.status === "success") {
+      displayStatus = "completed successfully";
+    } else if (data.status === "cancelled") {
+      displayStatus = "completed with some cancellations";
+    } else if (data.status === "failed") {
+      displayStatus = "completed with some failures";
+    } else if (data.status === "in-progress") {
+      displayStatus = "in progress";
```

Comment:
(same here) I see this pattern of string comparisons frequently in our `.js` files. Can we make this cleaner by using a `const object` (since JS doesn't have native enums)?